### PR TITLE
Extract RSpec Rails cops

### PR DESCRIPTION
**Pull request by ydah**

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -29,23 +29,18 @@ desc 'Build config/default.yml'
...
-  # of their relation with RuboCop RSpec.
-  rspec_cop_path = File.join('lib', 'rubocop', 'cop', 'rspec', 'base.rb')
+  require 'rubocop-rspec-rails'
+  require 'rubocop/rspec/rails/config_formatter'
+  require 'rubocop/rspec/rails/description_extractor'
+
+  glob = File.join('lib', 'rubocop', 'cop', 'rspec', 'rails', '*.rb')
   YARD::Tags::Library.define_tag('Cop Safety Information', :safety)
-  YARD.parse(Dir[glob].prepend(rspec_cop_path), [])
+  YARD.parse(Dir[glob], [])
```
Comment (bquorning) [inline Rakefile:38]: Why is this change needed?

File: config/default.yml [hunk 2]
```diff
@@ -1,12 +1,11 @@
...
-  DocumentationBaseURL: https://docs.rubocop.org/rubocop-rspec
-  Include: &1
+  Include:
     - "**/*_spec.rb"
     - "**/spec/**/*"
-  Language: &2
+  DocumentationBaseURL: https://docs.rubocop.org/rubocop-rspec-rails
+  Language:
     inherit_mode:
       merge:
```
Comment (bquorning) [inline config/default.yml]: Do we need this `Language` config still?

File: rubocop-rspec-rails.gemspec [hunk 3]
```diff
@@ -32,12 +29,11 @@ Gem::Specification.new do |spec|
...
-    'documentation_uri' => 'https://docs.rubocop.org/rubocop-rspec/',
+    'changelog_uri' => 'https://github.com/rubocop/rubocop-rspec_rails/blob/master/CHANGELOG.md',
+    'documentation_uri' => 'https://docs.rubocop.org/rubocop-rspec_rails/',
     'rubygems_mfa_required' => 'true'
   }
 
   spec.add_runtime_dependency 'rubocop', '~> 1.40'
-  spec.add_runtime_dependency 'rubocop-capybara', '~> 2.17'
-  spec.add_runtime_dependency 'rubocop-factory_bot', '~> 2.22'
+  spec.add_runtime_dependency 'rubocop-rspec', '~> 2.27.1'
```
Comment (bquorning) [inline rubocop-rspec-rails.gemspec]: I am not sure we want this line.

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -1,15 +1,18 @@
 # frozen_string_literal: true
 
 require 'rubocop'
-require 'rubocop/rspec/support'
+require 'rubocop/rspec/support' # `expect_offense` etc¥
```
Comment (bquorning) [inline spec/spec_helper.rb]:
```suggestion
require 'rubocop/rspec/support' # `expect_offense` etc
```

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -40,12 +43,12 @@ module SpecHelper
   # We should take their advice!
   config.raise_on_warning = true
 
-  config.include(ExpectOffense)
+  config.include RuboCop::RSpec::ExpectOffense
 
   config.include_context 'with default RSpec/Language config', :config
   config.include_context 'smoke test', type: :cop_spec
```
Comment (bquorning) [inline spec/spec_helper.rb]: I don’t think we need these two lines.

File: tasks/cops_documentation.rake [hunk 6]
```diff
@@ -13,61 +13,8 @@ end
 desc 'Generate docs of all cops departments'
 task generate_cops_documentation: :yard_for_generate_documentation do
   RuboCop::Cop::Registry.with_temporary_global do
```
Comment (bquorning) [inline tasks/cops_documentation.rake]: I am not sure we need the `with_temporary_global` block – rubocop-factory_bot doesn’t have it.

**Review by bquorning — COMMENTED** [review]
I tried diff'ing this code against a checkout of rubocop-factory_bot and rubocop-capybara, and it looks mostly ok. There are a few differences that I’ve tried to comment on.

File: Rakefile (hunk 1 above)
Comment (ydah) [reply Rakefile:38]: The base cop files `lib 'and' rubocop/cop/rspec/base.rb` are deleted because they are not needed, so are they needed?

File: rubocop-rspec-rails.gemspec (hunk 3 above)
Comment (ydah) [reply rubocop-rspec-rails.gemspec]: For example, https://github.com/rubocop/rubocop-rspec/blob/master/lib/rubocop/rspec/language.rb thinks that RuboCop RSpecRails should be usable. What do you think?

File: tasks/cops_documentation.rake (hunk 6 above)
Comment (ydah) [reply tasks/cops_documentation.rake]: Indeed. I removed it.

File: spec/spec_helper.rb (hunk 5 above)
Comment (ydah) [reply spec/spec_helper.rb]: RSpec Rails is also RSpec, so I thought it would be nice to have RSpec/Language loaded and smoke_test for RSpec. What do you think?

**Review by ydah — COMMENTED** [review]
I updated this PR.

File: README.md [hunk 7]
```diff
@@ -1,87 +1,79 @@
...
 ```
 
-### Code Climate
-
-rubocop-rspec is available on Code Climate as part of the rubocop engine. [Learn More](https://marketing.codeclimate.com/changelog/55a433bbe30ba00852000fac/).
-
 ## Documentation
 
-You can read more about RuboCop RSpec in its [official manual](https://docs.rubocop.org/rubocop-rspec).
+You can read more about RuboCop RSpec in its [official manual](https://docs.rubocop.org/rubocop-rspec_rails).
```
Comment (pirj) [inline README.md]: Rails?

File: README.md [hunk 8]
```diff
@@ -1,87 +1,79 @@
...
 
 ## The Cops
 
 All cops are located under
-[`lib/rubocop/cop/rspec`](lib/rubocop/cop/rspec), and contain
+[`lib/rubocop/cop/rspec_rails`](lib/rubocop/cop/rspec_rails), and contain
 examples/documentation.
 
 In your `.rubocop.yml`, you may treat the RSpec cops just like any other
```
Comment (pirj) [inline README.md]: Rails

File: docs/modules/ROOT/pages/index.adoc [hunk 9]
```diff
@@ -1,59 +1,14 @@
-= RuboCop RSpec
+= RuboCop RSpec Rails
 
-https://rspec.info/[RSpec]-specific analysis for your projects, as an extension to
+https://rspec.info/[RSpec Rails]-specific analysis for your projects, as an extension to
```
Comment (pirj) [inline docs/modules/ROOT/pages/index.adoc]:
```suggestion
https://rspec.info/features/6-1/rspec-rails/[RSpec Rails]-specific analysis for your projects, as an extension to
```

File: lib/rubocop-rspec_rails.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+require 'pathname'
+require 'yaml'
+
+require 'rubocop'
+require 'rubocop-rspec'
```
Comment (pirj) [inline lib/rubocop-rspec_rails.rb:7]: What if we remove this line?

File: lib/rubocop/cop/rspec/align_left_let_brace.rb [hunk 11]
```diff
@@ -1,49 +0,0 @@
-# frozen_string_literal: true
```
Comment (pirj) [inline lib/rubocop/cop/rspec/align_left_let_brace.rb:1]: The trick as I remember it was to wipe out files that won’t be present in the target state of the repo before the PR. Not remove them in the PR. Exceptions are lib/rubocop-rspec.rb -> lib/rubocop-rspec_rails.rb. At least this is how we did it for capybara https://github.com/rubocop-hq/rubocop-capybara/pull/1#issuecomment-1363270769

the difference is that in the final resulting repo we won’t keep much of the history of unrelated files (eg factory bot cops and docs).

File: rubocop-rspec-rails.gemspec [hunk 12]
```diff
@@ -1,25 +1,22 @@
 # frozen_string_literal: true
 
 $LOAD_PATH.unshift File.expand_path('lib', __dir__)
-require 'rubocop/rspec/version'
+require 'rubocop/rspec_rails/version'
 
 Gem::Specification.new do |spec|
-  spec.name = 'rubocop-rspec'
-  spec.summary = 'Code style checking for RSpec files'
+  spec.name = 'rubocop-rspec_rails'
```
Comment (pirj) [inline rubocop-rspec-rails.gemspec]: File name doesn’t match

File: lib/rubocop-rspec_rails.rb (hunk 10 above)
Comment (ydah) [reply lib/rubocop-rspec_rails.rb:7]: The following error occurs:

```
 ❯ bundle exec rake
rake aborted!
NameError: uninitialized constant RuboCop::Cop::RSpec (NameError)

      class AvoidSetupHook < ::RuboCop::Cop::RSpec::Base
                                           ^^^^^^^
/ydah/tmp/rubocop-rspec/lib/rubocop/cop/rspec_rails/avoid_setup_hook.rb:19:in `<module:RSpecRails>'
/ydah/tmp/rubocop-rspec/lib/rubocop/cop/rspec_rails/avoid_setup_hook.rb:5:in `<module:Cop>'
/ydah/tmp/rubocop-rspec/lib/rubocop/cop/rspec_rails/avoid_setup_hook.rb:4:in `<module:RuboCop>'
/ydah/tmp/rubocop-rspec/lib/rubocop/cop/rspec_rails/avoid_setup_hook.rb:3:in `<top (required)>'
/ydah/tmp/rubocop-rspec/lib/rubocop/cop/rspec_rails_cops.rb:3:in `require_relative'
/ydah/tmp/rubocop-rspec/lib/rubocop/cop/rspec_rails_cops.rb:3:in `<top (required)>'
/ydah/tmp/rubocop-rspec/lib/rubocop-rspec_rails.rb:8:in `require_relative'
/ydah/tmp/rubocop-rspec/lib/rubocop-rspec_rails.rb:8:in `<top (required)>'
tasks/cops_documentation.rake:4:in `<top (required)>'
/ydah/tmp/rubocop-rspec/Rakefile:19:in `load'
/ydah/tmp/rubocop-rspec/Rakefile:19:in `block in <top (required)>'
/ydah/tmp/rubocop-rspec/Rakefile:19:in `each'
/ydah/tmp/rubocop-rspec/Rakefile:19:in `<top (required)>'
/ydah/.rbenv/versions/3.3.0/bin/bundle:25:in `load'
/ydah/.rbenv/versions/3.3.0/bin/bundle:25:in `<main>'
(See full trace by running task with --trace)
```

File: lib/rubocop/cop/rspec/align_left_let_brace.rb (hunk 11 above)
Comment (ydah) [reply lib/rubocop/cop/rspec/align_left_let_brace.rb:1]: Oops, it seems that the deleted files were also included in the PR when extracting the FacotryBot division.
https://github.com/ydah/rubocop-factory_bot/pull/1

The steps at the time of extraction also remained below
https://github.com/rubocop/rubocop-factory_bot/pull/1#issuecomment-1507154570

File: rubocop-rspec-rails.gemspec (hunk 12 above)
Comment (ydah) [reply rubocop-rspec-rails.gemspec]: Oops, Indeed. Thank you!

File: lib/rubocop-rspec_rails.rb (hunk 10 above)
Comment (pirj) [reply lib/rubocop-rspec_rails.rb:7]: Got it. We need to understand what those cops need.
Even though it’s impossible to use rspec-rails without rspec, I don’t think we should follow the same with our cops. I’d preferred if this could be used separately from rubocop-rspec, won’t rely on rubocop-rspec releases, won’t break when things get fixed/broken in rubocop-rapec etc. This would mean we can’t keep the same base. What is the best way to proceed?

File: lib/rubocop/cop/rspec/align_left_let_brace.rb (hunk 11 above)
Comment (pirj) [reply lib/rubocop/cop/rspec/align_left_let_brace.rb:1]: Ouch, that would undermine the whole approach. We could just have copied the whole repo and made necessary changes to turn it into the result. But the problem is that we keep unrelated history.
Do we care? I don’t know. We keep the history of files that remain anyway, like lib/rubocop-<gemname>.rb, changelog, docs.

Even though the PR in your repo does seem to include deleted files, the other one, in rubocop namespace, doesn’t with a mere 40 files touched in the PR. Also if you take a look at the [repo before the PR](https://github.com/rubocop/rubocop-factory_bot/tree/812771eb33453e40f51c7a1cc993d7c8b6f5b33a/lib/rubocop/cop), there are only factory_bot cops there.
I suggest to follow the same approach.

Comment (ydah) [reply lib/rubocop/cop/rspec/align_left_let_brace.rb:1]: Maybe, I were probably taking care of the logs in some way. I'll try the previous procedure again this time and see if it's no problem.

File: Rakefile (hunk 1 above)
Comment (ydah) [reply Rakefile:38]: I put it back in https://github.com/ydah/rubocop-rspec_rails/pull/1/commits/e0cd330f08d3f9a015b770a07ccb9ba92248b4ad

File: rubocop-rspec-rails.gemspec (hunk 3 above)
Comment (ydah) [reply rubocop-rspec-rails.gemspec]: Removed in https://github.com/ydah/rubocop-rspec_rails/pull/1/commits/e0cd330f08d3f9a015b770a07ccb9ba92248b4ad

File: lib/rubocop-rspec_rails.rb (hunk 10 above)
Comment (ydah) [reply lib/rubocop-rspec_rails.rb:7]: I tried to get rid of the dependency in https://github.com/ydah/rubocop-rspec_rails/pull/1/commits/e0cd330f08d3f9a015b770a07ccb9ba92248b4ad What do you think?

File: rubocop-rspec_rails.gemspec [hunk 13]
```diff
@@ -34,5 +34,6 @@ Gem::Specification.new do |spec|
     'rubygems_mfa_required' => 'true'
   }
 
+  spec.add_runtime_dependency 'rack'
```
Comment (bquorning) [inline rubocop-rspec_rails.gemspec]: For the time being, we’ll probably have `rubocop-rspec_rails` being a runtime dependency of `rubocop-rspec`, just like `rubocop-capybara` and `rubocop-factory_bot` currently are: https://github.com/rubocop/rubocop-rspec/blob/447fa4e3d63e8998c8039ff7b9a2b216f61aa35c/rubocop-rspec.gemspec#L41-L42

And we don’t want `rack` to be a transitive runtime dependency of rubocop-rspec.

Comment (ydah) [reply rubocop-rspec_rails.gemspec]: Ah, I see. My apologize. I revert it.

**Comment by ydah** [comment]
I updated this PR.

**Comment by pirj** [comment]
Looks good!
I think the procedure was to:
1. Get the final snapshot using a PR like this
2. Build a new empty’ish repo with just the file set as the target
3. Get this git tool to clean history for files not present in the repo
4. Copy over the changes from the final snapshot and send a PR

Please correct me if I’m missing something

File: spec/spec_helper.rb [hunk 14]
```diff
@@ -18,7 +19,7 @@ module SpecHelper
 RSpec.configure do |config|
   # Set metadata so smoke tests are run on all cop specs
   config.define_derived_metadata(
-    file_path: %r{/spec/rubocop/cop/rspec/(?!mixin/)}
+    file_path: %r{/spec/rubocop/cop/rspec_rails/}
```
Comment (bquorning) [inline spec/spec_helper.rb]: rubocop-factory_bot [has this line instead](https://github.com/rubocop/rubocop-factory_bot/blob/fe2b31d038c197be211f74ba214879a58d31df80/spec/spec_helper.rb#L21):
```suggestion
    file_path: %r{/spec/rubocop/cop/}
```

File: lib/rubocop/rspec_rails/node.rb [hunk 15]
```diff
@@ -1,8 +1,8 @@
 # frozen_string_literal: true
 
 module RuboCop
-  module RSpec
-    # RuboCop RSpec specific extensions of RuboCop::AST::Node
+  module RSpecRails
+    # RuboCop RSpec Rails specific extensions of RuboCop::AST::Node
     module Node
```
Comment (bquorning) [inline lib/rubocop/rspec_rails/node.rb]: You can delete this file.

File: lib/rubocop-rspec_rails.rb [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+require_relative 'rubocop/rspec_rails/node'
+require_relative 'rubocop/rspec_rails/version'
+
+require_relative 'rubocop/cop/rspec_rails/base'
+require_relative 'rubocop/cop/rspec_rails_cops'
+
+project_root = File.join(__dir__, '..')
+RuboCop::ConfigLoader.inject_defaults!(project_root)
+
+RuboCop::AST::Node.include(RuboCop::RSpecRails::Node)
```
Comment (bquorning) [inline lib/rubocop-rspec_rails.rb]:
```suggestion
```

File: lib/rubocop-rspec_rails.rb [hunk 17]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+require 'pathname'
+require 'yaml'
+
+require 'rubocop'
+
+require_relative 'rubocop/rspec_rails/language'
+require_relative 'rubocop/rspec_rails/node'
```
Comment (bquorning) [inline lib/rubocop-rspec_rails.rb]:
```suggestion
```

File: lib/rubocop/rspec_rails/language.rb [hunk 18]
```diff
@@ -13,7 +13,6 @@ module RSpec
     # using the configured aliases.
     module Language
```
Comment (bquorning) [inline lib/rubocop/rspec_rails/language.rb]: Won’t it be too confusing to have both rubocop-rspec and rubocop-rspec_rails define a `Language` module. Would it be feasible to include `RuboCop::RSpec::Language` into the new Base class?

Or wait… maybe it’s just the `Language` configuration in config/default.yml I have objections to.

File: lib/rubocop/rspec_rails/description_extractor.rb [hunk 19]
```diff
@@ -32,8 +32,8 @@ def initialize(yardoc)
         # Test if the YARD code object documents a concrete rspec cop class
         #
         # @return [Boolean]
-        def rspec_cop?
-          cop_subclass? && !abstract? && rspec_cop_namespace?
+        def rspec_rails_cop?
+          cop_subclass? && !abstract? && rspec_rails_cop_namespace?
```
Comment (bquorning) [inline lib/rubocop/rspec_rails/description_extractor.rb]: I think the last part is not necessary.
```suggestion
          cop_subclass? && !abstract?
```

File: lib/rubocop/rspec_rails/description_extractor.rb [hunk 20]
```diff
@@ -32,8 +32,8 @@ def initialize(yardoc)
         # Test if the YARD code object documents a concrete rspec cop class
```
Comment (bquorning) [inline lib/rubocop/rspec_rails/description_extractor.rb]:
```suggestion
        # Test if the YARD code object documents a concrete cop class
```

**Review by bquorning — COMMENTED** [review]
rubocop-factory_bot does not have the Base class, and the Language module is included only in the cops where it is needed. I tried doing the same here in rubocop-rspec_rails, without success.

File: lib/rubocop/rspec_rails/language.rb (hunk 18 above)
Comment (ydah) [reply lib/rubocop/rspec_rails/language.rb]: I see. I was conversing with @pirj in the following thread, and I don't think we should depends on rubocop-rspec. duplicating Language is redundant, but I think it's necessary for each to be independent. I wonder if the reason it doesn't feel right is because the class name is RSpecRails::Language?

https://github.com/ydah/rubocop-rspec_rails/pull/1#discussion_r1520037079

File: lib/rubocop/rspec_rails/node.rb (hunk 15 above)
Comment (ydah) [reply lib/rubocop/rspec_rails/node.rb]: Thank you! I remove this file.

File: lib/rubocop/cop/rspec_rails/inferred_spec_type.rb [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+        alias on_numblock on_block
+
+        private
+
+        # @!method describe_with_type(node)
+        #   @param [RuboCop::AST::BlockNode] node
+        #   @return [RuboCop::AST::PairNode, nil]
+        def_node_matcher :describe_with_type, <<~PATTERN
+          (block
+            (send #rspec? #ExampleGroups.all
```
Comment (pirj) [inline lib/rubocop/cop/rspec_rails/inferred_spec_type.rb:80]: Ok, now I see where I was previously wrong with suggesting to untangle rubocop-rspec_rails from rubocop-rspec. The DSL config.

To avoid crutches and require our users to make head-tricks, I believe we would have to:
 - keep rubocop-rspec a runtime dependency in the new extension, with strict version constraints
 - add build jobs ensurng that it works fine with both the edge rubocop-rspec and the minimum required versions
 - … i had somethinng third in mind, but it has just slipped 🙈

Comment (pirj) [reply lib/rubocop/cop/rspec_rails/inferred_spec_type.rb:80]: We were lucky enough capybara and Factory_bot cops didn‘t need those

Comment (ydah) [reply lib/rubocop/cop/rspec_rails/inferred_spec_type.rb:80]: I see. Sure, if we add job to CI and fix the version, it looks good. Add rubocop-rspec to runtime dependency and add CI jobs on the edge of rubocop-rspec.

**Comment by ydah** [comment]
I updated this PR.

**Comment by ydah** [comment]
@bquorning @pirj 
As a result of the discussion at https://github.com/ydah/rubocop-rspec_rails/pull/1#discussion_r1523286089 , I've made the following changes, and I would appreciate your opinion.
* Add rubocop-rspec to runtime dependencies
  * Remove redundant files
* Add CI job to test rubocop-rspec edges

File: lib/rubocop/cop/rspec_rails/base.rb [hunk 22]
```diff
@@ -19,7 +18,7 @@ def self.inherited(subclass) # rubocop:disable Lint/MissingSuper
         # that have no other means of accessing the configuration.
         def on_new_investigation
           super
-          RuboCop::RSpec::Language.config = config['RSpec']['Language']
+          RuboCop::RSpec::Language.config = config['RSpecRails']['Language']
```
Comment (pirj) [inline lib/rubocop/cop/rspec_rails/base.rb]: Shouldn’t this be kept to pick up rhe conf as defined by the default conf or rubocop-rapec (and only optionally augmented by other gems and project config)?

File: lib/rubocop/rspec_rails/shared_contexts/default_rspec_language_config_context.rb [hunk 23]
```diff
@@ -1,15 +1,15 @@
 # frozen_string_literal: true
 
-RSpec.shared_context 'with default RSpec/Language config' do
+RSpec.shared_context 'with default RSpecRails/Language config' do
```
Comment (pirj) [inline lib/rubocop/rspec_rails/shared_contexts/default_rspec_language_config_context.rb]: I think we had a ticket to expose this in lib in rubocop-rspec. Can we reuse that shared example from there and remove it completely here?

File: lib/rubocop/rspec_rails/shared_contexts/default_rspec_language_config_context.rb [hunk 24]
```diff
@@ -1,15 +1,15 @@
...
   include_context 'config'
 
   # Deep duplication is needed to prevent config leakage between examples
   let(:other_cops) do
     default_language = RuboCop::ConfigLoader
-      .default_configuration['RSpec']['Language']
+      .default_configuration['RSpecRails']['Language']
```
Comment (pirj) [inline lib/rubocop/rspec_rails/shared_contexts/default_rspec_language_config_context.rb]: Where would this come from? We shouldn’t have any RSpecRails entry in the config

File: config/default.yml [hunk 25]
```diff
@@ -1,12 +1,11 @@
...
   Enabled: true
-  StyleGuideBaseURL: https://rspec.rubystyle.guide
-  DocumentationBaseURL: https://docs.rubocop.org/rubocop-rspec
-  Include: &1
+  DocumentationBaseURL: https://docs.rubocop.org/rubocop-rspec_rails
+  Include:
     - "**/*_spec.rb"
     - "**/spec/**/*"
-  Language: &2
+  Language:
```
Comment (pirj) [inline config/default.yml]: This needs to be removed. We don’t want our users (and gem authors who make their additions) to override it for both RSpec/Language and RSpecRails/Language

**Comment by pirj** [comment]
This is really close to readiness. Just one thing to reuse the config left. Thank you for the immense effort!

File: lib/rubocop/cop/rspec_rails/base.rb (hunk 22 above)
Comment (ydah) [reply lib/rubocop/cop/rspec_rails/base.rb]: Indeed! Thank you!

File: lib/rubocop/rspec_rails/shared_contexts/default_rspec_language_config_context.rb (hunk 24 above)
Comment (ydah) [reply lib/rubocop/rspec_rails/shared_contexts/default_rspec_language_config_context.rb]: Oops, I will remove it. Thank you.

File: config/default.yml (hunk 25 above)
Comment (ydah) [reply config/default.yml]: Indeed it is. I remove it!

**Comment by ydah** [comment]
@pirj Thank you for your review! I updated this PR.

**Comment by bquorning** [comment]
I’m sorry for not having had the capacity to review your PR again.

**Comment by ydah** [comment]
@bquorning Don't worry about it. Thank you as always 😄 !

**Review by pirj — APPROVED** [review]
Thank you!

File: rubocop-rspec_rails.gemspec [hunk 26]
```diff
@@ -32,12 +29,11 @@ Gem::Specification.new do |spec|
...
-    'documentation_uri' => 'https://docs.rubocop.org/rubocop-rspec/',
+    'changelog_uri' => 'https://github.com/rubocop/rubocop-rspec_rails/blob/master/CHANGELOG.md',
+    'documentation_uri' => 'https://docs.rubocop.org/rubocop-rspec_rails/',
     'rubygems_mfa_required' => 'true'
   }
 
   spec.add_runtime_dependency 'rubocop', '~> 1.40'
-  spec.add_runtime_dependency 'rubocop-capybara', '~> 2.17'
-  spec.add_runtime_dependency 'rubocop-factory_bot', '~> 2.22'
+  spec.add_runtime_dependency 'rubocop-rspec', '~> 2.27.1'
```
Comment (pirj) [inline rubocop-rspec_rails.gemspec]: Should we make it depend on 2.28.0 where all cops are extracted to prevent duplicates? I can’t recall how we did this in the past.
Point to github from Gemfile to ensure edge, release this gem first, then release rubocop-rspec 2.28.0, then change this gemfile to remove github source and point to rubygems.
I haven’t had my morning coffee yet, so please take this idea with a grain of salt.

Comment (ydah) [reply rubocop-rspec_rails.gemspec]: In the past we didn't need to worry about it, since rubocop-capybara and rubocop-factory_bot didn't rely on rubocop-rspec. Sorry. I want you to tell me, what kind of problem do I have if I don't depend on it after it's extracted?

Comment (pirj) [reply rubocop-rspec_rails.gemspec]: In theory, it would be possible to install rubocop-rspec_rails 2.28.0 with rubocop-rapec 2.27.1, where the extraction didn’t happen yet.
It could happen, but seems unrealistic.
Other cases would be like the one we’ve had when we wanted to yank a certain version? In some circumstances, bundler can prefer to downgrade rubocop-rspec to this combination of versions.
https://github.com/rubocop/rubocop-rspec/pull/1567

And, we should lick this to <3.0 here

Comment (ydah) [reply rubocop-rspec_rails.gemspec]: Thank you. Indeed it is. I update this PR https://github.com/ydah/rubocop-rspec_rails/pull/1/commits/1f412fd0016ac1d51d002db33f4fe8965b30fac9

**Comment by ydah** [comment]
@bquorning @pirj @Darhazer Now that all the problems have been solved, I will proceed with the extraction if there are no other problems.

File: rubocop-rspec_rails.gemspec (hunk 26 above)
Comment (pirj) [reply rubocop-rspec_rails.gemspec]: Ah, I see, the [version in edge is 2.27.1](https://github.com/rubocop/rubocop-rspec/blob/281a65bfefd8c01c152990000b95a836a834ef29/lib/rubocop/rspec/version.rb#L7), so it won’t work to point to it.

@bquorning @Darhazer do you foresee any issues with the current approach?

**Comment by ydah** [comment]
Okay, I'm going to proceed with the extraction.

**Comment by ydah** [comment]
Commands:
```
git clone --no-local rubocop-rspec rubocop-rspec_rails-hist
cd rubocop-rspec_rails-hist
git-filter-repo $( (cat ../rubocop-rspec_rails/rename | xargs -I_ echo '--path-rename _ ') )
git-filter-repo $( (cd ../rubocop-rspec_rails && git ls-files) | xargs -I_ echo '--path _ ')
git remote add origin git@github.com:rubocop/rubocop-rspec_rails.git
git push --set-upstream origin master
cd ../rubocop-rspec_rails
cp -r * ../rubocop-rspec_rails-hist
cp -r $( git ls-files | grep '^\.' | grep -v '\.github' ) ../rubocop-rspec_rails-hist
cp -r .github ../rubocop-rspec_rails-hist
cd -
git commit --all --message 'RSpec -> RSpecRails'
bundle install
bundle exec rake
git push origin master
```

rename:
```
docs/modules/ROOT/pages/cops_rspec_rails.adoc:docs/modules/ROOT/pages/cops_rspecrails.adoc
rubocop-rspec.gemspec:rubocop-rspec_rails.gemspec
spec/rubocop/cop/rspec/rails/avoid_setup_hook_spec.rb:spec/rubocop/cop/rspec_rails/avoid_setup_hook_spec.rb
spec/rubocop/cop/rspec/rails/have_http_status_spec.rb:spec/rubocop/cop/rspec_rails/have_http_status_spec.rb
spec/rubocop/cop/rspec/rails/http_status_spec.rb:spec/rubocop/cop/rspec_rails/http_status_spec.rb
spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:spec/rubocop/cop/rspec_rails/inferred_spec_type_spec.rb
spec/rubocop/cop/rspec/rails/minitest_assertions_spec.rb:spec/rubocop/cop/rspec_rails/minitest_assertions_spec.rb
spec/rubocop/cop/rspec/rails/negation_be_valid_spec.rb:spec/rubocop/cop/rspec_rails/negation_be_valid_spec.rb
spec/rubocop/cop/rspec/rails/travel_around_spec.rb:spec/rubocop/cop/rspec_rails/travel_around_spec.rb
lib/rubocop-rspec.rb:lib/rubocop-rspec_rails.rb
lib/rubocop/cop/rspec/rails/avoid_setup_hook.rb:lib/rubocop/cop/rspec_rails/avoid_setup_hook.rb
lib/rubocop/cop/rspec/rails/have_http_status.rb:lib/rubocop/cop/rspec_rails/have_http_status.rb
lib/rubocop/cop/rspec/rails/http_status.rb:lib/rubocop/cop/rspec_rails/http_status.rb
lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:lib/rubocop/cop/rspec_rails/inferred_spec_type.rb
lib/rubocop/cop/rspec/rails/minitest_assertions.rb:lib/rubocop/cop/rspec_rails/minitest_assertions.rb
lib/rubocop/cop/rspec/rails/negation_be_valid.rb:lib/rubocop/cop/rspec_rails/negation_be_valid.rb
lib/rubocop/cop/rspec/rails/travel_around.rb:lib/rubocop/cop/rspec_rails/travel_around.rb
lib/rubocop/cop/rspec_cops.rb:lib/rubocop/cop/rspec_rails_cops.rb
lib/rubocop/rspec/config_formatter.rb:lib/rubocop/rspec_rails/config_formatter.rb
lib/rubocop/rspec/description_extractor.rb:lib/rubocop/rspec_rails/description_extractor.rb
lib/rubocop/rspec/version.rb:lib/rubocop/rspec_rails/version.rb
```
